daq/report.py

```diff
@@ -246,33 +252,102 @@ def _write_test_tables(self):
         self._write_result_table()
         self._writeln()
 
-    def _write_category_table(self):
+    def _analyse_and_write_results(self):
+        """ Analyse the test results to determine if the device is a pass or fail
+            and identify possible issues (e.g. gone) and writes these to the report
+        """
         passes = True
+        gone = False
+
+        # Analyse results
+        for test_name, result_dict in self._results.items():
+            test_info = self._get_test_info(test_name)
+
+            if 'required' in test_info:
+                required_result = test_info['required']
+
+                # The device overall fails if any result is unexpected
+                if result_dict["result"] != required_result:
+                    passes = False
+
+                if result_dict["result"] == 'gone':
+                    gone = True
+
+        # Write Results
+        self._writeln('Overall device result %s' % ('PASS' if passes else 'FAIL'))
+        self._writeln()
+
+        if gone:
+            gone_message = ('**Some tests report as GONE. '
+                            'Please check for possible misconfiguration**')
+            self._writeln(gone_message)
+            self._writeln()
+
+    def _join_category_results(self, results):
+        """ Used to convert list of results into the pass/fail/skip format
+            for category table
+
+        Args:
+        results: List of results
+
+        Returns:
+        String in pass/fail/skip format
+        """
+        return '/'.join(str(value) for value in results)
+
+    def _write_category_table(self):
+        """ Write the first category and expected table
+        """
+
         rows = []
+        self._category_headers = self._CATEGORY_BASE_COLS + self._expected_headers
+
         for category in self._categories:
             total = 0
-            match = 0
+
+            results = [[0, 0, 0] for _ in range(len(self._expected_headers))]
+            result = self._NO_REQUIRED  # Overall category result
+
             for test_name, result_dict in self._results.items():
                 test_info = self._get_test_info(test_name)
                 category_name = test_info.get('category', self._DEFAULT_CATEGORY)
+
+                # all tests must have required in order to be counted
                 if category_name == category and 'required' in test_info:
-                    required_result = test_info['required']
                     total += 1
-                    if result_dict["result"] == required_result:
-                        match += 1
+
+                    expected_name = test_info.get('expected', self._DEFAULT_EXPECTED)
+                    expected_index = self._expected_headers.index(expected_name)
+
+                    # Put test results into right location in the result matrix
+                    if result_dict["result"] in ("pass", "info"):
```

Comment:
@noursaidi thanks for putting this together.
About the `info` result output, I think we agreed to remove it.
OK to do this in a separate PR though.

Reply:
@pisuke correct, we have issue #712 to cover the removal of info tests which requires requires more work and modification to various test modules. The reference to info tests would be removed as a part of those changes. Until then though, this ensures they are still continue to be accounted for as they are currently are